core/src/test/scala/org/apache/spark/storage/BlockManagerDecommissionSuite.scala

```diff
@@ -69,36 +84,64 @@ class BlockManagerDecommissionSuite extends SparkFunSuite with LocalSparkContext
     })
 
     // Cache the RDD lazily
-    sleepyRdd.persist()
+    if (persist) {
+      testRdd.persist()
+    }
 
     // Start the computation of RDD - this step will also cache the RDD
-    val asyncCount = sleepyRdd.countAsync()
+    val asyncCount = testRdd.countAsync()
 
     // Wait for the job to have started
     sem.acquire(1)
 
+    // Give Spark a tiny bit to start the tasks after the listener says hello
+    Thread.sleep(100)
+
     // Decommission one of the executor
     val sched = sc.schedulerBackend.asInstanceOf[StandaloneSchedulerBackend]
     val execs = sched.getExecutorIds()
-    assert(execs.size == 3, s"Expected 3 executors but found ${execs.size}")
+    assert(execs.size == numExecs, s"Expected ${numExecs} executors but found ${execs.size}")
     val execToDecommission = execs.head
+    logDebug(s"Decommissioning executor ${execToDecommission}")
     sched.decommissionExecutor(execToDecommission)
 
     // Wait for job to finish
-    val asyncCountResult = ThreadUtils.awaitResult(asyncCount, 3.seconds)
+    val asyncCountResult = ThreadUtils.awaitResult(asyncCount, 6.seconds)
     assert(asyncCountResult === 10)
     // All 10 tasks finished, so accum should have been increased 10 times
     assert(accum.value === 10)
 
     // All tasks should be successful, nothing should have failed
     sc.listenerBus.waitUntilEmpty()
-    assert(taskEndEvents.size === 10) // 10 mappers
+    if (shuffle) {
+      // 10 mappers & 10 reducers
+      assert(taskEndEvents.size === 20,
+        s"Expected 20 tasks got ${taskEndEvents.size} (${taskEndEvents})")
+    } else {
+      // 10 mappers
+      assert(taskEndEvents.size === 10,
+        s"Expected 10 tasks got ${taskEndEvents.size} (${taskEndEvents})")
```

Comment:
This test is quite sensitive right now.
I have seen a failed task because of "Launching a task while in decommissioned state." so it might be safer to count the successful tasks here, sth like:
 
```scala
   // 10 mappers which executed successfully   
   assert(taskEndEvents.count(_.reason == Success) === 10,
      s"Expected 10 tasks got ${taskEndEvents.size} (${taskEndEvents})")
```

And remove the `assert(taskEndEvents.map(_.reason).toSet === Set(Success))` below.

Reply:
Good catch!